File: src/algorithm_practice/Searching_And_Sorting/mergeSort/mergeSort.cpp
#include <iostream>
#include <climits>

using namespace std;

// My blog post on merging two sorted lists: https://chinocode.com/Merging-K-Sorted-Arrays-and-Linked-Lists/

/**
 * Merge Sort
 * Divide and conquer sorting routine
 * Time complexity: O(nlog(n))
 * Space complexity: O(n)
 *
 * The idea is simple. If we want to sort
 * an entire array, we first sort the left
 * half, then sort the right half, then merge
 * both arrays together. This yields a divide and
 * conquer algorithm because we first divide the
 * input in half, recursively solving each subproblem
 * and then merge the result. In short this means we'll
 * be merging all n elements once, n/2 elements twice,
 * n/4 four times, and so on. This gives us the following
 * sum: n + 2*(n/2) + 4*(n/4) + 8*(n/8) + ... + n*(n/n). It
 * should be clear that we're always doing n amount of work
 * at each level, and we're doing n amount of work exactly the
 * number of times we can split n in half. This number is log(n)
 * therefore the time complexity of this routine is O(nlog(n)).
 *
 * Merging two sorted contiguous arrays can be done in O(n) time.
 * Full documentation on this algorithm can be found in the above
 * linked blog post however we basically want to look at the first
 * values of each array and copy the smallest of the two into the
 * final array. We then need to make sure we're looking at the next
 * value of the array we took the smaller value from on our next iteration.
 * We can stop once we exhaust one array. Then we want to copy all of
 * the contents of the non-exhausted array into our final array. Below
 * I have two slightly different merge algorithms. They differ in the way
 * they exhaust one of the lists.
 *
 * Performing the merge operation in constant space turns
 * out to be very difficult. It is an area of research
 * still under study. Konrod's method of internal buffering
 * and block partitioning is one known way to do it. More information:
 * http://stackoverflow.com/questions/2126219/
 */

// Utility function to print array given size
void printArray(int arr[], int size) {
  cout << "{ ";

  for (int i = 0; i < size; ++i) {
    cout << arr[i];

    if (i < size - 1) {
      cout << ", ";
    }
  }

  cout << " }" << endl;
}

/**
 * This merge method will exhaust one array and then
 * copy the contents of the other to our final array.
 * We do this by going into a while loop while each
 * array has elements to contribute to our final array.
 * When we break out we simply run a different while
 * loop on the array that still has elements left to
 * contribute. This last while loop will copy all of
 * the elements from the non-exhausted array to the final
 * ouput array.
 */
void mergeNoSentinel(int list[], int l, int m, int r) {
  int n1 = m - l + 1;
  int n2 = r - m;

  int *L = new int[n1];
  int *R = new int[n2];
  for (int i = 0; i < n1; i++)
    L[i] = list[l + i];
  for (int j = 0; j < n2; j++)
    R[j] = list[m + 1 + j];

  int i = 0;
  int j = 0;
  int k = l;

  while (i < n1 && j < n2) list[k++] = (L[i] < R[j]) ? L[i++] : R[j++];

  while (i < n1) list[k++] = L[i++];

  while (j < n2) list[k++] = R[j++];

  delete [] L;
  delete [] R;
}

/**
 * This merge method adds a sentinel value
 * INT_MAX to the end of each sorted array.
 * Then for as many values that exist in each
 * sorted array not including our sentinel value
 * we take the smallest of the two elements appearing
 * at the beginning of each array. We then increment
 * the index value we used to get this smaller element.
 * Eventually when one array is exhausted we'll be looking
 * at the sentinel value INT_MAX from it, and this will
 * force us to always take values from the non-exhausted
 * array until it too becomes exhausted. At this point we've
 * used all of our elements and our loop will stop.
 */
void merge(int arr[], int l, int m, int r) {
  int n1 = m - l + 1;
  int n2 = r - m;

  int *a1 = new int[n1+1];
  int *a2 = new int[n2+1];

  for (int i = 0; i < n1; ++i)
    a1[i] = arr[l + i];
  for (int j = 0; j < n2; ++j)
    a2[j] = arr[m + 1 + j];

  a1[n1] = INT_MAX;
  a2[n2] = INT_MAX;

  int j = 0, k = 0;

  /**
   * One loop since we used sentinel
   * values designed to pull values from
   * the non-exhausted array once one
   * becomes exhausted.
   */
  for (int i = l; i <= r; ++i) {
    if (a1[j] < a2[k]) arr[i] = a1[j++];
    else arr[i] = a2[k++];
  }

  delete[] a1;
  delete[] a2;
}

void mergeSort(int list[], int l, int r){

  /**
   * We only want to sort the array if
   * the right index is greater than
   * the left index. If they are equal
   * we have a single element array which
   * is by definition sorted. To walk through
   * our next edge case, when an input array
   * contains two values i.e. when r = l+1, we
   * can see our m will be calculated as l after
   * floor division, thus splitting the array into
   * halves consisting of the first element and the
   * second element. Since this is a basic demonstration
   * of floor division, we can be sure that this will work
   * for all r > l.
   */
  if (r > l) {
    int m = (r - l)/2 + l; // integer overflow-safe way of doing int m = (l + r)/2;
    mergeSort(list, l, m);
    mergeSort(list, m+1, r);
    merge(list, l, m, r);
  }

}

int main() {
  int a[] = {5, 4, 3, 2, 1};
  mergeSort(a, 0, 4);
  printArray(a, 5);

  return 0;
}

File: src/algorithm_practice/Searching_And_Sorting/mergeSort/mergeSort.py
#=======================================================================
#  Statement:
#  Given a disordered list of integers (or any other items),
#  rearrange the integers in natural order.
#
#  Sample Input: [8,5,3,1,9,6,0,7,4,2,5]
#
#  Sample Output: [0,1,2,3,4,5,5,6,7,8,9]
#
#  Time Complexity of Solution:
#  Best = Average = Worst = O(nlog(n)).
#
#  Approach:
#   Merge sort is a divide and conquer algorithm. In the divide and
#   conquer paradigm, a problem is broken into pieces where each piece
#   still retains all the properties of the larger problem -- except
#   its size. To solve the original problem, each piece is solved
#   individually; then the pieces are merged back together.
#
#   For illustration, imagine needing to sort an array of 200 elements
#   using selection sort. Since selection sort takes O(n^2), it would
#   take about 40,000 time units to sort the array. Now imagine
#   splitting the array into ten equal pieces and sorting each piece
#   individually still using selection sort. Now it would take 400
#   time units to sort each piece; for a grand total of 10400 = 4000.
#   Once each piece is sorted, merging them back together would take
#   about 200 time units; for a grand total of 200+4000 = 4,200.
#   Clearly 4,200 is an impressive improvement over 40,000. Now
#   imagine greater. Imagine splitting the original array into
#   groups of two and then sorting them. In the end, it would take about
#   1,000 time units to sort the array. That's how merge sort works.
#
#  NOTE to the Python experts:
#     While it might seem more "Pythonic" to take such approach as
#
#         mid = len(aList) / 2
#         left = mergesort(aList[:mid])
#         right = mergesort(aList[mid:])
#
#     That approach take too much memory for creating sublists.
#======================================================================= 
def mergesort( aList ):
  _mergesort( aList, 0, len( aList ) - 1 )
 
 
def _mergesort( aList, first, last ):
  # break problem into smaller structurally identical pieces
  mid = ( first + last ) / 2
  if first < last:
    _mergesort( aList, first, mid )
    _mergesort( aList, mid + 1, last )
 
  # merge solved pieces to get solution to original problem
  a, f, l = 0, first, mid + 1
  tmp = [None] * ( last - first + 1 )
 
  while f <= mid and l <= last:
    if aList[f] < aList[l] :
      tmp[a] = aList[f]
      f += 1
    else:
      tmp[a] = aList[l]
      l += 1
    a += 1
 
  if f <= mid :
    tmp[a:] = aList[f:mid + 1]
 
  if l <= last:
    tmp[a:] = aList[l:last + 1]
 
  a = 0
  while first <= last:
    aList[first] = tmp[a]
    first += 1
    a += 1